### ISSUE TITLE
Implementing memory-augmented recurrent networks efficiently

### ISSUE THREAD

**Issue by rizar** [Theano/Theano#5566]
When implementing a memory-augmented network (for example, [Neural Turing Machine](https://arxiv.org/abs/1410.5401) or a simpler architecture like an [LSTM with attention over its past hidden states](https://arxiv.org/abs/1601.06733)) using Theano one has to deal with huge memory consumption. Here I discuss a way to implement them efficiently, but seems like this would require new features on Theano side.

In general, such architectures involve a computation of the following form:

```
s[0] = the initial state of the controller
M[0] = the initial state of the memory
for i in range(1, T + 1):
   r[i] = READ(M[i - 1], s[i - 1])
   s[i] = NEXT_STATE(s[i - 1], x[i], r[i])
   M[i] = WRITE(M[i - 1], s[i])
```

The difference between s[i] and M[i] is usually that M[i] is much larger and that the WRITE operation is typically simpler than NEXT_STATE. Moreover, quite often the write operation can be "undone", i.e. there exists UNWRITE such that UNWRITE(WRITE(M, S), S) = M.

With a naive implementation of back-propagation all the states s[i] and memory states M[i] will be reused in the backward pass, which means they will all be kept in memory until the backward pass is done. It's okay to keep s[i], but M[i] are too big and memory consumption becomes a big issue. A solution is to use instead the recomputed memory states Mr[i], where the recomputation is done in the reverse order:

```
Mr[T] = the final memory state
for i in range(T - 1, 0, -1):
   Mr[i] = UNWRITE(Mr[i + 1], s[i])
```

If in the backward pass Mr[i] are used instead of M[i], they could be recomputed on-the-go and garbage collected = memory efficient implementation.

We brainstormed yesterday with @abergeron about how this could be implemented and ran out of ideas. Even if we do not consider the fact that such architectures are always implemented with `theano.scan`, it's not clear how to force `theano.grad` to use replace `M[i]` with `Mr[i]` in the backward pass. It's even worse with `theano.scan`, as `Mr[i]` would have to be added as an additional output to the backpropagation scan. 

In general, this would be a super cool feature helpful for many people if Theano could do such things. Is there a chance that support for that can be added with relatively few changes?

@caglar , you might be interested.

**Comment by nouiz** [comment]
Could we pass a known_grad parameter to the call to theano.grad() done
inside scan.grad (scan.L_op in fact)?

Mostly, if we do it without scan, would that work? If so, we can find a way
to pass the information to scan as a second step.

Maybe you can define in the known_grad the grad for M computed from Mr
instead from M[i-1].

**Comment by rizar** [comment]
I agree that perhaps we should first find a solution without scan. You suggested 

> Maybe you can define in the known_grad the grad for M computed from Mr
instead from M[i-1].

The gradient dr[i]/dM[i-1] will typically only depend on S[i-1] and not on M[i-1]. The hard part is the gradient dr[i]/dS[i-1] which depends on M[i-1]. If we only consider one iteration, then indeed, we could first compute dL/dr[i], then provide this gradient as `known_grads` to a `theano.grad` call that differentiates `READ(Mr[i-1], S[i-1])`. But in practice there will a lot of iterations, and that's where things get complicated.

One possible solution is to have a `replace` argument in `theano.grad` that contains maps forward pass (M[i]) variables into their substitutes that are computed in the reverse order (Mr[i])

**Comment by khaotik** [comment]
> One possible solution is to have a replace argument in theano.grad that contains maps forward pass

What if you use `OpFromGraph(grad_overrides=...)` to define a completely custom gradient routine?

```python
grad_op = OpFromGraph( [<inputs>], [<scan>] )
fwd_op = OpFromGraph([M[-1], ...], [<scan>], grad_overrides=grad_op)
```

I suppose, if it possible to make sure `M[-1]` is *not* done inplace, then the garbage collector would work
